### PR TITLE
Org: Automatic archiving and deletion of tickets.

### PR DESCRIPTION
**Pull request by cyrillkuettel**
## Commit message

Org: Automatic archiving and deletion of tickets.

TYPE: Feature
LINK: OGC-58

## Checklist

- [x] I have performed a self-review of my code
- [x] I considered adding a reviewer
- [x] I have updated the PO files
- [x] I made changes/features for both org and town6
- [x] I have tested my code thoroughly by hand
- [x] I have added tests for my changes/features
- [x] I am using freezegun for tests depending on date and times

### CODE REVIEW

File: src/onegov/org/forms/settings.py
Comment (Daverball) [inline src/onegov/org/forms/settings.py:1147]: Why is this a function? As far as I can tell there's nothing dynamic being generated here.

I would move this into a global constant, e.g. `TIMESPAN_CHOICES`. 

Also it's worth noting that this doesn't actually work because both `SelectField` and `RadioField` default their `coerce` argument to `str`, i.e. all the values are turned into strings.

I think the sanest way to solve this would be to pass `coerce=int` and use integers for the choices, then you can use `0` for disabled. Then you can either override `process_obj`/`populate_obj` as you have or simplify things by leaving the setting as an integer and convert it to a timedelta in the cronjob.

File: src/onegov/org/forms/settings.py
Comment (Daverball) [inline src/onegov/org/forms/settings.py:1167]: You don't actually need to call `super().process_obj(model)` here, since you are manually setting the only fields in the form.

But by adressing the above comment you may be able to get rid of this entirely anyways.

File: src/onegov/org/models/ticket.py
Comment (Daverball) [inline src/onegov/org/models/ticket.py:382]:
```suggestion
        exists = self.session.query(Reservation).filter(
            Reservation.start > func.now()
        ).exists()

        return self.session.query(exists).scalar()
```
An exists subquery is much more efficient than a query limited to 1 record. (It also avoids the ORM overhead)

But also, shouldn't you filter the reservations to the ones attached to the ticket? Otherwise you will only ever be able to delete tickets if no future reservations exist at all.

File: src/onegov/org/models/ticket.py
Comment (Daverball) [inline src/onegov/org/models/ticket.py:426]:
```suggestion
        for reservation in self.reservations or ():
            self.session.delete(reservation)
```

File: src/onegov/org/models/ticket.py
Comment (Daverball) [inline src/onegov/org/models/ticket.py:695]: Have we fully thought this through, should past events actually ever be deleted?

File: src/onegov/org/views/ticket.py
Comment (Daverball) [inline src/onegov/org/views/ticket.py:1107]:
```suggestion
        if not getattr(ticket.handler, 'ticket_deletable', True):
```
`hasattr` + attribute access can be simplified to `getattr` with default.

Although I think the default should be flipped, it doesn't really seem safe to me to delete a ticket when the handler doesn't specify whether it should be deleted or not.

Comment (Daverball) [reply src/onegov/org/views/ticket.py:1107]: But also I would honestly just add `ticket_deletable` with a sane default to the base handler class, then you don't need `getattr`/`hasattr` at all.

File: src/onegov/org/views/ticket.py
Comment (Daverball) [inline src/onegov/org/views/ticket.py:1115]: Same here, if this is added to the base handler, then we don't need this `hasattr` check.

File: src/onegov/org/views/ticket.py
Comment (Daverball) [inline src/onegov/org/views/ticket.py:1139]:
```suggestion
```
I think you can get rid of this, `cascade` should take care of this for you.

**Review by Daverball — CHANGES_REQUESTED** [review]
Looks good for the most part, but there are still some bugs and rough edges here and there.

File: src/onegov/agency/models/ticket.py
Comment (Daverball) [reply src/onegov/agency/models/ticket.py:59]: Actually I would go a step further and just add this to `Handler`, maybe even get rid of `TicketDeletionMixin`. Are there any handlers that don't use this mixin?

File: src/onegov/org/forms/settings.py
Comment (cyrillkuettel) [reply src/onegov/org/forms/settings.py:1147]: ah yes probably better than handling the strings. This itched me anyway, having to depend on string parsing of datetime for this critical part of the code.

File: src/onegov/org/models/ticket.py
Comment (cyrillkuettel) [reply src/onegov/org/models/ticket.py:382]: Very good catch here, this got removed when I refactored. the previous version I had was using `self.reservations` which is what we probably want here.

File: src/onegov/org/models/ticket.py
Comment (cyrillkuettel) [reply src/onegov/org/models/ticket.py:695]: It would only apply to those which have no future occurrence, else it would not be `deletable`. 

`future_occurrences. (https://github.com/OneGov/onegov-cloud/blob/76cd47f707f42b0f41f0a8d4e369fba985c5a966/src/onegov/org/models/ticket.py#L705C34-L705C34)

But maybe we should reconsider it anyway...

File: src/onegov/org/views/ticket.py
Comment (cyrillkuettel) [reply src/onegov/org/views/ticket.py:1139]: Probably yes, I'll have a test for that.

File: src/onegov/org/forms/settings.py
Comment (cyrillkuettel) [reply src/onegov/org/forms/settings.py:1147]: And  it's a function because we would have to duplicate it two times for the both fields of this form....

File: src/onegov/org/models/ticket.py
Comment (Daverball) [reply src/onegov/org/models/ticket.py:695]: Yes, but it still seems weird, since events are public information, tickets are not.

File: src/onegov/org/views/ticket.py
Comment (cyrillkuettel) [reply src/onegov/org/views/ticket.py:1115]: Seems obvious in hindsight. There's not really a reason not to add it to the base handler.

File: src/onegov/org/models/ticket.py
Comment (cyrillkuettel) [reply src/onegov/org/models/ticket.py:382]: I still can't believe I missed that. Might have to add a test for that

File: src/onegov/agency/models/ticket.py
Comment (cyrillkuettel) [reply src/onegov/agency/models/ticket.py:59]: The mixin is everywhere. I just wanted to be "extra safe" like what if someone happens to write a handler in the future without `TicketDeletionMixin`. That was the original idea. But moving it up the inheritance hierarchy seems like the way to go.

File: src/onegov/org/models/ticket.py
Comment (cyrillkuettel) [reply src/onegov/org/models/ticket.py:695]: Yeah. I think we should not do that. What's also just struck me just now is that `view_event` makes a query for the associated ticket. If it does not exist, it will open a new one !?  This might result in unexpected things happening. (https://github.com/OneGov/onegov-cloud/blob/9af704438ccae09b3a416704b4fcc02d99c19d23/src/onegov/org/views/event.py#L256)

This is all getting a bit convoluted.

File: src/onegov/org/forms/settings.py
Comment (Daverball) [reply src/onegov/org/forms/settings.py:1147]: you don't need a function to use the same choices in multiple fields. We have other fields that share common choices, such as `WEEKDAYS`, those are defined in a global variable as well.

Comment (cyrillkuettel) [reply src/onegov/org/forms/settings.py:1147]: ah right, it only needs computed once.

File: src/onegov/org/cronjobs.py
Comment (cyrillkuettel) [reply src/onegov/org/cronjobs.py]: Done!

File: src/onegov/feriennet/models/activity.py
Comment (Daverball) [inline src/onegov/feriennet/models/activity.py:143]: Reversing the order of these conditions might be slightly more efficient, since I'm not sure if `self.activity` is eagerly loaded or not.

File: src/onegov/org/cronjobs.py
Comment (Daverball) [inline src/onegov/org/cronjobs.py:52]: Is the idea here that we don't enable this part of the cronjob until a later date?

File: src/onegov/org/cronjobs.py
Comment (Daverball) [inline src/onegov/org/cronjobs.py:488]: This is not correct, a method is always truthy, you actually need to call the method:
```suggestion
        if getattr(ticket.handler, 'has_future_reservation', lambda: False)():
```

File: src/onegov/org/models/organisation.py
Comment (Daverball) [inline src/onegov/org/models/organisation.py:184]: If you change these to have a default value then you don't need a custom `process_obj` method in the form.
```suggestion
    auto_archive_timespan = meta_property(default=0)
    auto_delete_timespan = meta_property(default=0)
```

File: src/onegov/org/models/ticket.py
Comment (Daverball) [inline src/onegov/org/models/ticket.py:340]: It's probably not worth to make the query a `cached_property`. I would change this to `property`

File: src/onegov/org/models/ticket.py
Comment (Daverball) [inline src/onegov/org/models/ticket.py:353]: Maybe we should make this a `cached_property`, unless you're confident we only call this once per request. This will also mean changing all the calls to simple attribute accesses.

File: src/onegov/org/models/ticket.py
Comment (Daverball) [inline src/onegov/org/models/ticket.py:405]: I'm not sure how `super()` interacts with properties, but I think this should still be `super().ticket_deletable` without the call.

File: src/onegov/org/templates/cautionary_form.pt
Comment (Daverball) [inline src/onegov/org/templates/cautionary_form.pt:1]: Wouldn't a `request.alert(msg)` in the view be sufficient here? Is it actually worth adding a new template for this?

File: src/onegov/ticket/handler.py
Comment (Daverball) [inline src/onegov/ticket/handler.py:232]: This assertion suggest to me that we might want `ticket_deletable` to change to a `cached_property` in some cases, otherwise we do the work of figuring out if the ticket is deletable twice.

**Review by Daverball — CHANGES_REQUESTED** [review]
Looks good for the most part, but there are still a couple of small problems.

File: src/onegov/org/cronjobs.py
Comment (Daverball) [reply src/onegov/org/cronjobs.py:488]: If you decide to change this to a `cached_property`, like suggested further below, then this doesn't need to change.

File: src/onegov/org/cronjobs.py
Comment (cyrillkuettel) [reply src/onegov/org/cronjobs.py:52]: This needs to be removed, was just for local testing. I have configured both jobs to run in the night (e.g. once every 24 hours)

File: src/onegov/org/cronjobs.py
Comment (cyrillkuettel) [reply src/onegov/org/cronjobs.py:488]: Good catch here. If we change `has_future_reservation` to be cached_property, this ` lambda: False)()`  would not be necessary ?

Comment (Daverball) [reply src/onegov/org/cronjobs.py:488]: correct

File: src/onegov/org/templates/cautionary_form.pt
Comment (cyrillkuettel) [reply src/onegov/org/templates/cautionary_form.pt:1]: yes, actually that's right. New template is overkill

File: src/onegov/org/cronjobs.py
Comment (cyrillkuettel) [reply src/onegov/org/cronjobs.py:488]: That would also be a good case for a linter. I'm actually kind of surprised

Comment (Daverball) [reply src/onegov/org/cronjobs.py:488]: mypy would catch this, but as long as this doesn't have type annotations it will not.

Comment (Daverball) [reply src/onegov/org/cronjobs.py:488]: Actually nevermind the `getattr` just inherently makes this difficult to statically analyze.

Comment (cyrillkuettel) [reply src/onegov/org/cronjobs.py:488]: I was contemplating using `isinstance` anyway. Do you know if there's not much of a performance difference between that and `getattr` ?

File: src/onegov/ticket/handler.py
Comment (cyrillkuettel) [reply src/onegov/ticket/handler.py:232]: I'm kind of questioning if the assertion is even worth it to write. That was there before in  `prepare_delete_ticket`, I did not add this

Comment (Daverball) [reply src/onegov/ticket/handler.py:232]: It's a pretty good sanity check to make sure this only gets called for tickets that are actually deletable.

File: src/onegov/org/cronjobs.py
Comment (Daverball) [reply src/onegov/org/cronjobs.py]: shouldn't be a big hit, there's some things like runtime checkable protocols which would be slow in an `isinstance` check and there are other classes that implement their own `__instancecheck__` which has a potential to be slow, but this shouldn't apply here.

**Comment by cyrillkuettel** [comment]
Anything I missed, are we all set? @Daverball

**Comment by Daverball** [comment]
@cyrillkuettel You will need to do another merge to resolve conflicts, otherwise I think we're good to go, but I will double check after the merge just in case.

**Review by Daverball — APPROVED** [review]

**Pull request merged**
